File: neetcode/go/125.valid-palindrome.go
/*
 * @lc app=leetcode id=125 lang=golang
 *
 * [125] Valid Palindrome
 */

// @lc code=start
import (
	str "strings"
	r "regexp"
)

func isPalindrome(s string) bool {
	alnum := r.MustCompile("^[a-z0-9]*$")

	isAlnum := func(char rune) bool {
		return alnum.MatchString(string(char))
	}

	filter := func(ss string, predicate func(rune) bool) string {
		var sb str.Builder
		for _, s := range ss {
			if predicate(s) {
				sb.WriteRune(s)
			}
		}
		return sb.String()
	}

	s = str.ToLower(s)
	s = filter(s, isAlnum)
	leftPtr := 0
	rightPtr := len(s) - 1
	for leftPtr < rightPtr {
		if s[leftPtr] == s[rightPtr] {
			leftPtr++
			rightPtr--
			continue
		} else {
			return false
		}
	}
	return true
}

// @lc code=end

File: neetcode/125.valid-palindrome.py
#
# @lc app=leetcode id=125 lang=python3
#
# [125] Valid Palindrome
#

# @lc code=start
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = ''.join(list(
            filter(
                lambda char: char.isalnum(),
                s
            )
        )).lower()
        left_ptr = 0
        right_ptr = len(s) - 1
        while left_ptr < right_ptr:
            if s[left_ptr] == s[right_ptr]:
                left_ptr += 1
                right_ptr -= 1
                continue
            else:
                return False
        return True
        
# @lc code=end

